sandbox/modules/identity/src/main/java/org/opensearch/identity/rest/user/UserService.java

```diff
@@ -0,0 +1,219 @@
+        }
+
+        // hash generated from provided plain-text password
+        if (password != null) {
+            // TODO: discuss if we are going to allow hash to be passed as request data instead of plain text password
+            userToBeCreated.setHash(Hasher.hash(password.toCharArray()));
+        }
+
+        // TODO: check if this is absolutely required
+        internalUsersConfiguration.remove(username);
+
+        // Create or update the user
+        internalUsersConfiguration.putCObject(username, userToBeCreated);
+
+        // save the changes to identity index, propagate change to other nodes and reload in-memory configuration
+        saveAndUpdateConfiguration(username, this.nodeClient, CType.INTERNALUSERS, internalUsersConfiguration, listener);
+    }
+
+    /**
+     * Load data for a given CType
+     * @param config CType whose data is to be loaded in-memory
+     * @return configuration loaded with given CType data
+     */
+    protected final SecurityDynamicConfiguration<?> load(final CType config) {
+        SecurityDynamicConfiguration<?> loaded = this.configurationRepository.getConfigurationsFromIndex(Collections.singleton(config))
+            .get(config)
+            .deepClone();
+        return loaded;
+    }
+
+    protected CType getConfigName() {
+        return CType.INTERNALUSERS;
+    }
+
+    /**
+     * Check if identity index exists in cluster
+     * @return true if exists, false otherwise
+     */
+    protected boolean ensureIndexExists() {
+        if (!this.clusterService.state().metadata().hasConcreteIndex(this.identityIndex)) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Persist changes to CType configuration in index, propagates this change to other nodes and reload in-memory cache
+     * @param username of the user to be persisted in the index
+     * @param client to execute index update request
+     * @param cType Config Type to be reloaded
+     * @param configuration Data to be persisted in index
+     * @param listener on which to send response once execution completes
+     */
+    protected void saveAndUpdateConfiguration(
+        final String username,
+        final Client client,
+        final CType cType,
+        final SecurityDynamicConfiguration<?> configuration,
+        ActionListener<CreateUserResponse> listener
+    ) {
+        // TODO: Future scope: see if this method can be generalized and extracted to another class
+
+        // Listener for responding once index update completes
+        final ActionListener<IndexResponse> actionListener = new OnSucessActionListener<>() {
+            @Override
+            public void onResponse(IndexResponse indexResponse) {
+                CreateUserResponseInfo responseInfo = new CreateUserResponseInfo(true, username);
+                CreateUserResponse response = new CreateUserResponse(unmodifiableList(asList(responseInfo)));
+
+                listener.onResponse(response);
```

Comment:
This looks fine to me, but I wanted to leave a comment showing the universe of replies that are currently in the security plugin for a comparable action: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/dlic/rest/api/AbstractApiAction.java#L481-L520

These responses give back different status codes for different scenarios. A batched API may get confusing if one of the items in the batch is created and another one updates.

Reply:
My understanding is that this onResponse returns successful only when index operation was successful, otherwise onFailure would be called. 

However to you point of update vs create, I'm returning a more generic message indicating it was successful. I initially had thought about returning a specific message with each success response, but that makes thing complicated unnecessarily. Here's how:
1. Modify CreateUserResponseInfo to accept a 3rd argument, say `message`
2. Pass an additional argument in `saveAndUpdateConfig(..., boolean userExisted)`, and update the call on line 132
3. Modify this `onResponse` implementation to set correct message in CreateUserResponseInfo constructor:
```java
...
String operationResponse = userExisted ? " updated successfully" : " created successfully";
String message = username + operationResponse;
CreateUserResponseInfo responseInfo = new CreateUserResponseInfo(true, username, message);
...
```

If we agree that this should be added, I can modify the code and the tests. This also means that we will support Updating a user via PUT in addition to POST.